### DISCUSSION TITLE
Realtime UR connection

### DISCUSSION THREAD

**Discussion by filipejsbrandao** [visose/Robots#306] [Q&A] [answered]
I'm trying to get the robot actual TCP position from a UR robot from grasshopper using robots. I have seen some previous posts here which use the URReal class which is deprecated. So I tried to implement it using the URRealTimeDataExchange class. I managed to connect to the robot and get some values in return, but all the values are always 0. 

For instance, the code below returns a array of 6 values in the TCPPosition output: [0,0,0,0,0,0]
I have to mention that I currently don't have access to the robot, so I'm testing this on a VM. Yet, it should be the same. What am I missing here?

`private void RunScript(
	string robotIP,
	bool getPosition,
	ref object TCPPosition,
	ref object LogMessages)
    {
        // Output variables
        TCPPosition = null;
        List<string> logs = new List<string>();

        // Check if the connection should be triggered
        if (getPosition)
        {
            try
            {
                // List of variables to request from the robot
                List<string> variables = new List<string> { "actual_TCP_pose" };

                // Create a new instance of the URRealTimeDataExchange class
                using (var urRTDE = new Robots.URRealTimeDataExchange(robotIP, variables))
                {
                    // Read data from the robot
                    urRTDE.ReadOutputs();

                    // Retrieve the TCP pose data
                    foreach (var output in urRTDE.Outputs)
                    {
                        if (output.Name == "actual_TCP_pose")
                        {
                            TCPPosition = output.Values; // Store the TCP position values
                            break;
                        }
                    }

                    // Add log messages
                    logs.AddRange(urRTDE.Log);
                }
            }
            catch (Exception ex)
            {
                // Handle exceptions and log errors
                logs.Add($"Error: {ex.Message}");
            }
        }
        else
        {
            logs.Add("Connection not triggered.");
        }

        // Output the logs
        LogMessages = logs;
    }`

**Comment by visose** [comment]
I would change this so that the URRealTimeDataExchange instance is a field rather than a local variable inside RunScript. Initialize it only once and  run the component on a loop with a timer, without reinitializing it on every call to RunScript.

**Reply by filipejsbrandao** [reply, marked as answer]
Thanks for the advice. I have done as you suggested but I'm still only getting [0,0,0,0,0,0] as TCP position. Here is how the code looks:

`public class Script_Instance : GH_ScriptInstance
{
    // Persistent fields
    private URRealTimeDataExchange urRTDE;  // Persistent RTDE connection
    private bool isConnected = false;      // Connection status
    private List<double> lastTCPPosition;  // Store last successful TCP position
    private List<string> logs = new List<string>(); // Log messages

    private void RunScript(string robotIP, bool connect, bool getUpdatedPosition, ref object TCPPosition, ref object LogMessages)
    {
        // Output variables
        TCPPosition = lastTCPPosition; // Default to the last successful position
        logs.Clear();                  // Reset logs for each run

        // Handle connection logic
        if (connect)
        {
            if (!isConnected)
            {
                try
                {
                    // Initialize connection if not already connected
                    List<string> variables = new List<string> { "actual_TCP_pose" };
                    urRTDE = new URRealTimeDataExchange(robotIP, variables);
                    isConnected = true;
                    logs.Add($"Connected to robot at {robotIP}");
                }
                catch (Exception ex)
                {
                    logs.Add($"Error connecting to robot: {ex.Message}");
                }
            }

            // If connected and an update is requested, fetch the TCP position
            if (isConnected && getUpdatedPosition)
            {
                try
                {
                    urRTDE.ReadOutputs();

                    // Retrieve the `actual_TCP_pose` data
                    foreach (var output in urRTDE.Outputs)
                    {
                        if (output.Name == "actual_TCP_pose")
                        {
                            lastTCPPosition = new List<double>(output.Values); // Update stored position
                            TCPPosition = lastTCPPosition;                    // Output updated position
                            logs.Add("TCP position updated successfully.");
                            break;
                        }
                    }

                    // Log messages from URRealTimeDataExchange
                    logs.AddRange(urRTDE.Log);
                }
                catch (Exception ex)
                {
                    logs.Add($"Error reading robot data: {ex.Message}");
                }
            }
        }
        else
        {
            // Disconnect the robot if `connect` is false
            if (isConnected)
            {
                urRTDE?.Dispose();
                urRTDE = null;
                isConnected = false;
                logs.Add("Disconnected from robot.");
            }
            else
            {
                logs.Add("Robot already disconnected.");
            }
        }

        // Output the logs
        LogMessages = logs;
    }

    // Destructor for proper cleanup
    ~Script_Instance()
    {
        urRTDE?.Dispose();
    }
}`

Thanks

**Reply by filipejsbrandao** [reply]
I think I found out what the problem is. If I send a program using the Robots Remote Connection component I have to reinitialise the connection to be able to get the values. So I imagine the remote connection component also uses RTDE to send programs to the robot..
Thanks!

**Reply by 1999-ali** [reply]
Hello,

I'm currently working on a very similar project using Rhino 8 with Grasshopper. I've successfully managed to upload commands to the UR5e robot—the upload part is functioning well. However, my main focus now is on receiving real-time feedback from the robot, such as TCP position, voltage, current, and other important data.

I attempted to follow the open-source code provided in this repository: [GH_Robots_URRealtimeFeedback](https://github.com/robin-gdwl/GH_Robots_URRealtimeFeedback.git), but unfortunately, I couldn't get it to work on my setup.

While exploring alternatives, I came across your post and I’m now considering building a custom solution if needed. Real-time feedback is critical for my application, and ideally, I’d like to receive updates every 10 ms for optimal performance.

One challenge I'm currently facing is that I’m not sure where to find the URRealTimeDataExchange class in the interface. I searched the Grasshopper canvas but couldn't locate it. Could you kindly provide some guidance or help regarding this issue?

Thank you in advance for your support!

**Reply by KonradJuenger** [reply]
hi @1999-ali just an heads up, anything under 30ms in grasshopper is almost impossible, maybe with a very minimal script. even though the GH trigger tries to run every _10ms_ or less, it will only trigger as quickly as one iteration through the canvas. With some complex logic you will have larger delays, even worse the updates are jittery. I would recommend separating real-time feedback functions from Gh as much as possible.

**Reply by filipejsbrandao** [reply]
My application did not demand 10ms. I was working around 100 to 200ms. Even so there was a state machine with a few if conditions to solve the problems @KonradJuenger just mentioned. There are also other things I bumped into. Some tools also use the RTDE interface, in my case the VGC10 gripper. My application was running for a few weeks, and I got a few instances where this would cause the robot to have an emergency stop.

**Reply by 1999-ali** [reply]
Thank you very much for your help! I'd really like to explore your solution — perhaps I can build upon it or adapt it to my needs.

At the moment, even a start with a 100–200 ms delay is something I can work with, and from there, I’ll see if it truly fits my requirements.

Would it be possible for you to share your source code or provide more detailed information? I’d really appreciate any insights you can offer.

Kind regards,
Ame

**Comment by 1999-ali** [comment]
Thank you very much for your help! I'd really like to explore your solution
— perhaps I can build upon it or adapt it to my needs.

At the moment, even a start with a 100–200 ms delay is something I can work
with, and from there, I’ll see if it truly fits my requirements.

Would it be possible for you to share your source code or provide more
detailed information? I’d really appreciate any insights you can offer.

Kind regards,
Ame
